### ISSUE TITLE
Disable backface culling when VR is initialized

### ISSUE THREAD

**Issue by cpinter** [KitwareMedical/SlicerVirtualReality#46]
Clipping occurs in VR very often, i.e. when moving your head close to a model. In this case, if backface culling is enabled (as is by default), then you can see through the model although there is a back side of it in front of you.

I propose changing the backface culling to false in the default model display node, and also in all existing model display nodes in the scene when VR is initialized.

Also we could add an Application settings panel where this could be the first option.
@jcfr @pieper @lassoan what do you think?

**Comment by jcfr** [comment]
Should back face culling be disabled only in object represented in VR world or in both regular 3D view and VR ?

**Comment by pieper** [comment]
I think only disabling backface culling VR by default would be best, since it could be a performance hit on slower graphics cards.  It could make sense to have an button to turn it on or off on all models in the Models module.

We should also be able to use the BackfaceProperty option on the actors so that it would be visually clear when you are seeing the inside of a model.  I haven't used this but I think it would be helpful.

**Comment by adamrankin** [comment]
There is some discussion on the internet about the performance impact of backface culling, and it seems that it's situational. For scenes with large amounts of polygons in a relatively small screen space area, back-face culling has a high impact. This might be the case if someone is zoomed out to view a complex brain structure. Zoomed in, the performance impact of backface culling would be minimal.

I suspect that it would be find to blanket disable backface culling when VR is enabled, and then devise a more effective/complicated solution when performance barriers are hit.

**Comment by lassoan** [comment]
OpenGL2 backend made vtkPolyData rendering about a magnitude faster (have not profiled it, it's just an estimate), so probably even 2x slower rendering due to backface culling would be still way faster than OpenGL1 rendering backend.

I've been thinking about disabling backface culling by default to reduce user confusion in various use cases other than virtual reality (e.g., when transforms are applied to a model that turns them inside out). 

If we disable backface culling by default in the display node constructor, users can still put a couple of lines in their slicerrc file to change the default on a specific computer.

> We should also be able to use the BackfaceProperty option on the actors so that it would be visually clear when you are seeing the inside of a model.

Is your concern that users would be confused because they could not tell if they see a front or back face? Would you automatically create backface property by modifying frontface property? What would you change? Adding some subtle pattern could be nice, but I don't know if VTK can do it. Tuning the color (e.g., decrease saturation), showing mesh lines, etc. would be quite arbitrary.

**Comment by jcfr** [comment]
I am curious to known what are the thoughts of @martinken

**Comment by pieper** [comment]
>> We should also be able to use the BackfaceProperty option on the actors so that it would be visually clear when you are seeing the inside of a model.
> 
> Is your concern that users would be confused because they could not tell if they see a front or back face? Would you automatically create backface property by modifying frontface property? What would you change? Adding some subtle pattern could be nice, but I don't know if VTK can do it. Tuning the color (e.g., decrease saturation), showing mesh lines, etc. would be quite arbitrary.

Yes, I think a subtle pattern would be nice.  This should be trivial in a fragment shader if the right hooks exist - I agree it may not be supported currently.  Otherwise yes, just making it darker would be a reasonable cue.  The best thing depends on the use case, so leaving it programmable is best.

The other issue is opacity - if you turn off backface culling and don't use depth peeling things get even more confusing.  With depth peeling I'd argue you usually want culling on.  Traditionally culling + transparency handled a reasonable subset of situations, even if it's not ideal.

**Comment by cpinter** [comment]
I tried to turn depth peeling on/off and it seems to have no effect in VR, only in the regular 3D view. Not sure why.

**Comment by lassoan** [comment]
> The other issue is opacity - if you turn off backface culling and don't use depth peeling things get even more confusing. With depth peeling I'd argue you usually want culling on. Traditionally culling + transparency handled a reasonable subset of situations, even if it's not ideal.

Depth peeling was broken in VTK for a long time. Now it works well and with depth peeling you actually want to see all sides of transparent models. Without that, intersection of back side of models is simply not correctly displayed.

**Comment by cpinter** [comment]
How about this:

I add an "Optimize scene for VR" button on the toolbar and in the module, which does certain changes that improve VR experience. Currently:
* Turn off backface culling for all models and make default model display node also disable backface culling
* Turn off slice intersection visibility for all models and segmentations (apply in default display nodes)

Later we can add options for these and maybe also auto-optimize on VR startup in application settings, but for now a button will be also helpful.

Setting backface properties does not seem too useful to me, because when we see the backface of a model/segmentation in VR, then it is always after user camera interaction, not like in a static view or even in a regular 3D view where camera interaction is less central. The user goes close enough to the object so that the inside is visible, and when moves back, the outside is visible only. So it's quite obvious when we're seeing the inside of something. So I'd not add this now, but we can consider it if needed.
